### PR TITLE
fix(ln): don't allow same preimage for sale twice

### PR DESCRIPTION
**Pull request by elsirion**
Mitigation for #1397.

TODO:
- [x] Unit test

### CODE REVIEW

File: modules/fedimint-ln-common/src/lib.rs [hunk 1]
```diff
@@ -365,6 +365,8 @@ pub enum LightningError {
     ZeroOutput,
     #[error("Offer contains invalid threshold-encrypted data")]
     InvalidEncryptedPreimage,
+    #[error("Offer contains a ciphertext that has already been usec")]
```
Comment (m1sterc001guy) [inline modules/fedimint-ln-common/src/lib.rs:368]: Nit: typo

**Review by m1sterc001guy — APPROVED** [review]
Looks good to me. Could you explain why this is a mitigation and not a full fix? Is it because a gateway could collude with the guardians to post the encrypted preimage twice?

**Comment by elsirion** [comment]
> why this is a mitigation and not a full fix

We use the threshold encryption scheme as a blackbox and assume some minimal set of guarantees like keeping the plaintext secret from an adversary. Something we don't know about the scheme is if there is a way for an adversary, given a ciphertext `c=Enc(p)`, to construct a `c'` that decrypts to a `p'=Dec(c')` such that the adversary can infer `p` from `p'`.

@joschisan already has a candidate scheme to fully fix the problem and also simplify the LN logic quite a bit at the same time (get rid of offers). But it still needs to be proven secure and implemented, which will probably take a bit.

**Comment by justinmoon** [comment]
> @joschisan already has a candidate scheme

Looking forward to this!!!

**Pull request merged**